### PR TITLE
Editable config

### PR DESCRIPTION
**Pull request by tama**
- Configuration values can be set per server (added commands to read a value and set it)
- Automatic cleaning of raid commands after the delay set in `mod.clean_cmd_after` configuration value (#24)

### CODE REVIEW

**Comment by nicolasblaye** [comment]
Can you add some test? It shouldn't take that long :)

File: mod_bot/conf.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import os
+import json
+import pprint
+
+def read_config(path):
+    print("reading configuration")
+    if os.path.exists(path) is False:
```
Comment (nicolasblaye) [inline mod_bot/conf.py]: maybe we should fail if we can't read a conf? Or set default values?

File: mod_bot/conf.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        return {}
+
+    with open(path, 'r') as f:
+        conf = json.load(f)
+    load_guilds_config(conf)
+    return conf
+
+def load_guilds_config(conf):
+    if "filepath" in conf and "listen_to" in conf:
+        for gid in list(conf["listen_to"]):
```
Comment (nicolasblaye) [inline mod_bot/conf.py]: I don't like abbreviated name because you have to think to read them, code should be more fluid

File: mod_bot/conf.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+def set_key(path, conf, gid, value):
+    if has_key('filepath', conf) is False:
+        print("No 'filepath' found in global config file, can't continue")
+        return
+
+    gconf_path = '{0}/{1}/config.json'.format(conf['filepath'], gid)
+    cur_config = read_config(gconf_path)
+    parts = path.split('.')
+    ptr = cur_config
+    for part in parts[:-1]:
```
Comment (nicolasblaye) [inline mod_bot/conf.py]: hard to follow

File: mod_bot/conf.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    # Update values so that the next read returns the new value
+    with open(gconf_path, 'w') as f:
+        json.dump(cur_config, f)
+    
+    load_guild_config(gid, conf)
+
+def dump(obj):
+    pprint.pprint(obj, indent=4)
+
+if __name__ == "__main__":
```
Comment (nicolasblaye) [inline mod_bot/conf.py]: useful?

File: mod_bot/main.py [hunk 5]
```diff
@@ -67,21 +76,39 @@ def add_minutes(hstart, mstart, mn):
             save("muted", muted_users)
             return
 
-    if str(message.guild.id) == '322379168048349185' and message.author.name in muted_users:
-        if int(time.time()) < int(muted_users[message.author.name]):
-            print("muted until {0} (current {1})".format(muted_users[message.author.name], int(time.time()))) 
+    # Mod commands
+    roles = discord.utils.get(message.guild.members, name=message.author.name).roles
+    mod_roles = get(gid, 'mod.roles', conf).split(';')
+    is_mod = False
```
Comment (nicolasblaye) [inline mod_bot/main.py]: make that into a function. this could latter go to admin.py file?

File: mod_bot/main.py [hunk 6]
```diff
@@ -67,21 +76,39 @@ def add_minutes(hstart, mstart, mn):
...
             await message.delete()
             return
         else:
-            del(muted_users[message.author.name])
+            del(muted_users[author])
             save("muted", muted_users)
 
-    guild_id = message.channel.guild.id
-    if is_listen_to(guild_id) and conf["listen_to"][str(guild_id)]["channel"] == str(message.channel.id) and message.author.name != 'modbot':
+    if is_listen_to(gid) and get(None, 'listen_to.{0}.channel'.format(gid), conf) == cid and author != 'modbot':
```
Comment (nicolasblaye) [inline mod_bot/main.py]: not a fan of the None as a value (functional programmer here :D )

File: mod_bot/conf.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        if part not in ptr:
+            # Invalid path
+            #print('Invalid path at {0}'.format(part))
+            return None
+        ptr = ptr[part]
+    return ptr
+
+def set_key(path, conf, gid, value):
+    if has_key('filepath', conf) is False:
+        print("No 'filepath' found in global config file, can't continue")
```
Comment (nicolasblaye) [inline mod_bot/conf.py]: this should send a message back to the user

File: mod_bot/conf.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import os
+import json
+import pprint
+
```
Comment (nicolasblaye) [inline mod_bot/conf.py]: Maybe config could be our first object? It would stored the internal config instead of passing it in each methods that needs to read the conf

Comment (tama) [reply mod_bot/conf.py]: i'm not a fan of "global variables", I think it's better to use DI (dependency injection) and therefore pass the configuration object in every method who needs it

File: mod_bot/conf.py (hunk 4 above)
Comment (tama) [reply mod_bot/conf.py]: Yup, for manual testing

File: mod_bot/conf.py (hunk 3 above)
Comment (tama) [reply mod_bot/conf.py]: Is it ? I'm not finding it hard to follow, just think of it as a pointer that moves inside the structure following the different parts of the path given

File: mod_bot/conf.py (hunk 2 above)
Comment (tama) [reply mod_bot/conf.py]: Given the context ("load_guilds_config"), I think the name is explicit enough

File: mod_bot/conf.py (hunk 1 above)
Comment (tama) [reply mod_bot/conf.py]: Where should we store the default values ? 
- In a configuration file ? And what if that configuration file cannot be read for whatever reason ?
- Inside the code ? Not a fan of this solution as we'd need to update it everytime some mandatory configuration is added or removed

I can make it fail if it can't read the file because it does not exist, but I prefer the "empty" configuration (where every "get" afterwards will not return anything) rather than having to handle the exception afterwards
